intro_to_ada.rst

```diff
@@ -1904,23 +1904,747 @@ Abstract data types
+    end Generic_Array_Pkg;
+
+Because we have a procedure declaration, we need a package body for the
+procedure implementation. Here, we haven't changed the previous algorithm
+-- we're simply moving the existing code into the new package body:
+
+.. code-block:: ada
+
+    package body Generic_Array_Pkg is
+       procedure Reverse_Array (X : in out Array_T) is
+       begin
+          for I in X'First .. (X'Last + X'First) / 2 loop
+             declare
+                Tmp     : T;
+                X_Left  : T renames X (I);
+                X_Right : T renames X (X'Last + X'First - I);
+             begin
+                Tmp     := X_Left;
+                X_Left  := X_Right;
+                X_Right := Tmp;
+             end;
+          end loop;
+       end Reverse_Array;
+    end Generic_Array_Pkg;
+
+In the test application, we just need to instantiate the
+``Generic_Array_Pkg`` package and make use of the array type (``Array_T``)
+and the procedure (``Reverse_Array``):
+
+.. code-block:: ada
+
+    with Ada.Text_IO;
+    use  Ada.Text_IO;
+
+    with Generic_Array_Pkg;
+
+    procedure Test_Reverse_Colors_Pkg is
+       type Color is (Black, Red, Green, Blue, White);
+
+       package Color_Pkg is new Generic_Array_Pkg (T => Color, Index => Integer);
+
+       My_Colors : Color_Pkg.Array_T (1 .. 5) := (Black, Red, Green, Blue, White);
+    begin
+       for C of My_Colors loop
+          Put_Line ("My_Color: " & Color'Image (C));
+       end loop;
+
+       New_Line;
+       Put_Line ("Reversing My_Color...");
+       New_Line;
+       Color_Pkg.Reverse_Array (My_Colors);
+
+       for C of My_Colors loop
+          Put_Line ("My_Color: " & Color'Image (C));
+       end loop;
+
+    end Test_Reverse_Colors_Pkg;
+
+
+Abstracting the test application
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In the previous examples, we've focused only on abstracting the reversing
+algorithm. However, we could have decided to also abstract our little
+test application. This could be useful if we, for example, decide to
+test other procedures that change elements of an array.
+
+In order to achieve this, we have to abstract quite a few elements. We
+will therefore declare the following formal parameters:
+
+    - ``S``: the string containing the array name
+
+    - an instance of the ``Generic_Array_Pkg`` package (which was
+      implemented in the previous section)
+
+    - a function ``Image`` that converts an element of type ``T`` to a
+      string
+
+    - a procedure ``Pkg_Test`` that performs some operation on the array
+
+Note that ``Image`` and ``Pkg_Test`` are examples of formal subprograms.
+Also, note that ``S`` is an example of a formal object.
+
+This is a version of the test application that makes use of the generic
+``Perform_Test`` procedure:
+
+.. code-block:: ada
+
+    with Ada.Text_IO;
+    use  Ada.Text_IO;
+
+    with Generic_Array_Pkg;
+
+    procedure Test_Reverse_Colors_Pkg is
+
+       generic
+          S : in String;
```

Comment:
`in` annotations are optional, and it is a style error in GNAT's coding style to add them, I'd rather we follow this style convention.

Reply:
Sure